scheduled_messages: Add keyboard navigation.

shameondev opened the pull request:
The main goal of this pull request is to make the keyboard navigation created for the drafts modal reusable, so we can use it in the scheduled messages modal as well.

The first commit refactors the keyboard functions in the `drafts.js` module for reusability, in preparation for implementing keyboard navigation in the message scheduling modal. The `modals_handle_events` function has been made reusable by extracting context-related variables and using a more neutral term "item" instead of "draft". The function also now accepts a `context` parameter to work in different modal contexts. 

The second commit adds keyboard navigation to the scheduled messages modal using the `modals_handle_events` function from the `drafts.js` module, along with some helpful functions from the same module.

<details>
<summary>Screenshots:</summary>
<br>

![image](https://github.com/zulip/zulip/assets/53193850/ad4ed8f5-1980-476a-a662-e0bb9057ab59)
![image](https://github.com/zulip/zulip/assets/53193850/b5074879-6a96-40d4-88ba-2dcd82aaee4f)


</details>

Fixes: #25181.

shameondev commented:
Overall, I'm feeling good about the work that's been done. However, I'm unsure if the patterns I chose are good enough. Do you think we should keep all the keyboard handling functions in `drafts.js`? Also, I'm not sure if the way I passed the `context` variable is the best approach. If you have any ideas or suggestions, I would appreciate them!

alya commented:
Works for me in manual testing!

Let's re-add the keyboard shortcuts to the tooltips in the same PR?

alya commented:
Not marked for integration review, as in addition to the comment above, looks like there's also a test to fix.

shameondev commented:
> Let's re-add the keyboard shortcuts to the tooltips in the same PR?

Sure, added them!

alya commented:
Could you post a screenshot for the tooltips, please?

shameondev commented:
@alya, I've fixed the failing test and added the shortcut hints to the tooltips!

shameondev commented:
Added screenshots!

timabbott commented on web/src/drafts.js:
```diff
@@ -689,12 +689,12 @@ export function launch() {
 
     open_overlay();
-    const first_element_id = [...formatted_narrow_drafts, ...formatted_other_drafts][0].draft_id;
+    const first_element_id = [...formatted_narrow_drafts, ...formatted_other_drafts][0]?.draft_id;
```
What's the story with this line? It seems like it might belong in the previous commit.

shameondev replied:
Yeah, that definitely belongs to the first commit, where I tried to make `set_initial_element` reusable. The optional chaining was added after the failed test, and it seems reasonable since there may be cases where there are no drafts or first element.

timabbott commented on web/src/scheduled_messages.js:
```diff
@@ -324,6 +360,10 @@ export function initialize(scheduled_messages_params) {
+
+    $("body").on("focus", ".overlay-message-info-box", (e) => {
+        drafts.activate_element(e.target, keyboard_handling_context);
+    });
```
The changes in this file all belong in `scheduled_messages_overlay_ui.js`, since they are implementation used for that overlay only.

shameondev replied:
Moved the changes into `scheduled_messages_overlay_ui.js`!

timabbott commented:
The first commit looks good. I'd like to see a new second commit that moves all the newly shared functions from `drafts.js` into a new `messages_overlay_ui.js`, and has `drafts.js` call into that, before the third commit (replacing the second commit you have here) having the scheduled messages code read that.

I also noticed a bug while working on this, which is that probably the scheduled messages UI should not have a click handler for clicking on a previously scheduled messages unscheduling it and moving it to the compose box; clicking on the message body (and not the edit/delete pencils) should just do nothing. Can you add a (I guess 4th) commit adjusting that?

@alya FYI we may also want a compose banner that gets displayed when you unschedule a message, since right now we don't have explicit feedback when you do that.

shameondev commented:
Thank you for the review, Tim!

I have addressed all of the feedback you provided, which includes:
1. Adding a commit that moves newly shared functions to a new module called `messages_overlay_ui.js`.
2. Adjusting the click handlers for the message body of the scheduled messages, so that clicking on them does nothing.

This PR is now ready for the next round of review!

alya commented:
Tooltips screenshots look good. Just to check, are we doing the smart thing of showing Enter vs. Return depending on your OS?

timabbott commented:
Merged, dropping the last commit, thanks for doing this @shameondev!
